test/matrix_utils.cc

```diff
@@ -0,0 +1,408 @@
+        tileNb = [nb](int64_t j) {
+            // for non-uniform tile size
+            return (j % 2 != 0 ? nb*2 : nb);
+        };
+    }
+    else {
+        // NB. we let BaseMatrix truncate the final tile length
+        // TrapezoidMatrix only takes 1 function for both dimensions (of different sizes)
+        tileNb = [nb](int64_t j) {
+            return nb;
+        };
+    }
+
+    auto tileRank = slate::func::process_2d_grid( grid_order, p, q );
+    int num_devices_ = blas::get_device_count();
+    auto tileDevice = slate::func::device_1d_grid( slate::GridOrder( dev_dist ),
+                                                   p, num_devices_ );
+
+    // Setup matrix to test SLATE with
+    if (origin != slate::Origin::ScaLAPACK) {
+        // SLATE allocates CPU or GPU tiles.
+        slate::Target origin_target = origin2target( origin );
+        if (nonuniform_nb || dev_dist == slate::Dist::Col) {
+            matrix.A = construct_irregular( tileNb, tileRank, tileDevice );
+        }
+        else {
+            matrix.A = construct_regular( nb, grid_order, p, q );
+        }
+        matrix.A.insertLocalTiles( origin_target );
+    }
+    else {
+        assert( !nonuniform_nb );
+        assert( dev_dist == slate::Dist::Row );
+        // Create SLATE matrix from the ScaLAPACK layouts
+        matrix.A_data.resize( matrix.lld * matrix.nloc );
+        matrix.A = construct_scalapack( &matrix.A_data[0], matrix.lld, nb,
+                                        grid_order, p, q );
+    }
+
+    // Setup reference matrix
+    if (ref_matrix) {
+        if (nonuniform_nb && nonuniform_ref) {
+            matrix.Aref = construct_irregular( tileNb, tileRank, tileDevice );
+            matrix.Aref.insertLocalTiles( slate::Target::Host );
+        }
+        else {
+            matrix.Aref_data.resize( matrix.lld * matrix.nloc );
+            matrix.Aref = construct_scalapack( &matrix.Aref_data[0], matrix.lld, nb,
+                                           grid_order, p, q );
+        }
+    }
+
+    return matrix;
+}
+
+// -----------------------------------------------------------------------------
+/// Allocates a Matrix<scalar_t> and optionally a reference version for testing.
+///
+/// @param[in] ref_matrix
+///     Whether to allocate a reference matrix
+///
+/// @param[in] nonuniform_ref
+///     If params.nonuniform_nb(), whether to also allocate the reference matrix
+///     with non-uniform tiles.
+///
+/// @param[in] m
+///     The number of rows
+///
+/// @param[in] n
+///     The number of columns
+///
+/// @param[in] params
+///     The test params object which contains many of the key parameters
+///
+template <typename scalar_t>
+TestMatrix<slate::Matrix<scalar_t>> allocate_test_Matrix(
+    bool ref_matrix, bool nonuniform_ref, int64_t m, int64_t n, Params& params)
+{
+    auto construct_irregular = [&] (nb_func_t tileNb,
+                                    dist_func_t tileRank, dist_func_t tileDevice)
```

Comment:
Meyers says to avoid default capture modes (item 31).
1. Capture by reference `[&]` can leave dangling references.
2. Capture by value `[=]` can mislead that you don't have dangling references. But for member vars, `this` pointer is captured, which can dangle.

In this case, all the lambdas go out-of-scope in this function, so they don't outlive this object.